examples/step-75/step-75.cc

```diff
@@ -0,0 +1,1575 @@
+
+
+
+  // @sect3{The <code>Preconditioner</code> class template}
+
+  // Preconditioner for the equation system.
+  template <typename VectorType>
+  class PreconditionerBase
+  {
+  public:
+    virtual void vmult(VectorType &dst, const VectorType &src) const = 0;
+  };
+
+  template <typename VectorType>
+  class PreconditionerAMG : public PreconditionerBase<VectorType>
+  {
+  public:
+    template <typename Operator>
+    PreconditionerAMG(const Operator &system_matrix)
+    {
+      LA::MPI::PreconditionAMG::AdditionalData data;
+      data.elliptic              = true;
+      data.higher_order_elements = true;
+      preconditioner.initialize(system_matrix.get_system_matrix(), data);
+    }
+
+    void vmult(VectorType &dst, const VectorType &src) const override
+    {
+      preconditioner.vmult(dst, src);
+    }
+
+  private:
+    LA::MPI::PreconditionAMG preconditioner;
+  };
+
+
+
+  // @sect3{The <code>AdaptationStrategy</code> class template}
+
+  // Different strategies to perform hp adaptation.
+  template <int dim>
+  class AdaptationStrategy
+  {
+  public:
+    AdaptationStrategy(const ParameterHandler &                   params,
+                       DoFHandler<dim> &                          dof_handler,
+                       parallel::distributed::Triangulation<dim> &triangulation)
+      : params(params)
+      , dof_handler(dof_handler)
+      , triangulation(triangulation)
+    {
+      for (unsigned int degree = min_degree; degree <= max_degree; ++degree)
+        face_quadrature_collection.push_back(QGauss<dim - 1>(degree + 1));
+    };
+
+    virtual void
+    adapt_resolution(const LinearAlgebra::distributed::Vector<double>
+                       &locally_relevant_solution);
+
+  protected:
+    void flag_adaptation(const LinearAlgebra::distributed::Vector<double>
+                           &locally_relevant_solution);
+    virtual void decide_hp(const LinearAlgebra::distributed::Vector<double>
+                             &locally_relevant_solution);
+    void         limit_levels();
+    virtual void execute_refinement();
+
+    const ParameterHandler &params;
+    const unsigned int      min_level = 5, max_level = dim <= 2 ? 10 : 8;
+    const unsigned int      min_degree = 2, max_degree = dim <= 2 ? 7 : 5;
+
+
+    DoFHandler<dim> &                          dof_handler;
+    parallel::distributed::Triangulation<dim> &triangulation;
+
+    hp::QCollection<dim - 1> face_quadrature_collection;
+
+    Vector<float> estimated_error_per_cell;
+  };
+
+
+
+  template <int dim>
+  void AdaptationStrategy<dim>::adapt_resolution(
+    const LinearAlgebra::distributed::Vector<double> &locally_relevant_solution)
+  {
+    flag_adaptation(locally_relevant_solution);
+    decide_hp(locally_relevant_solution);
+    limit_levels();
+    execute_refinement();
+  }
+
+
+
+  template <int dim>
+  void AdaptationStrategy<dim>::flag_adaptation(
+    const LinearAlgebra::distributed::Vector<double> &locally_relevant_solution)
```

Comment:
You use `AdaptationStrategy` both as a base class (=interface) and to implement a specific strategy. That seems wrong. Why not just have one interface base class and put all specific code into derived classes?

Reply:
Agreed.

However, there will be functions which are duplicates of each other. Maybe I'll introduce an anonymous or internal namespace instead which will feature these functions, and each class will simply call them instead of inheriting them from the mother class.